dace/codegen/targets/fpga.py

```diff
@@ -369,17 +371,51 @@ def generate_state(self, sdfg: dace.SDFG, state: dace.SDFGState,
 
             kernel_host_stream.write(f"""\
 DACE_EXPORTED void {host_function_name}({', '.join(kernel_args_opencl)}) {{
-      hlslib::ocl::Program program = __state->fpga_context->Get().CurrentlyLoadedProgram();"""
-                                     )
+      hlslib::ocl::Program program = __state->fpga_context->Get().CurrentlyLoadedProgram();\
+""")
             # Create a vector to collect all events that are being generated to allow
             # waiting before exiting this state
-            kernel_host_stream.write(f"std::vector<cl::Event> all_events;", )
+            kernel_host_stream.write("std::vector<cl::Event> all_events;")
 
             # Kernels invocations
             kernel_host_stream.write(state_host_body_stream.getvalue())
 
             # Wait for all events
-            kernel_host_stream.write(" cl::Event::waitForEvents(all_events);")
+            kernel_host_stream.write("cl::Event::waitForEvents(all_events);")
+
+            # Instrumentation
+            if state.instrument == dtypes.InstrumentationType.FPGA:
+                kernel_host_stream.write("""
+// Begin FPGA kernel runtime instrumentation
+cl_ulong first_start = std::numeric_limits<unsigned long int>::max();
+cl_ulong last_end = std::numeric_limits<unsigned long int>::min();""")
+                for i, sg in enumerate(all_subgraphs):
+                    module_name = self._module_name(sg, state)
+                    if Config.get_bool("instrumentation", "print_fpga_runtime"):
+                        print_str = f"""
+    const double elapsed = 1e-9 * (event_end - event_start);
+    std::cout << "Kernel \\"{module_name}\\" executed in " << elapsed << " seconds.\\n";\
+"""
+                    else:
+                        print_str = ""
+                    kernel_host_stream.write(f"""\
+{{
+    cl_ulong event_start = 0;
+    cl_ulong event_end = 0;
+    cl::Event const &event = all_events[{i}];
+    event.getProfilingInfo(CL_PROFILING_COMMAND_START, &event_start);
+    event.getProfilingInfo(CL_PROFILING_COMMAND_END, &event_end);
+    if (event_start < first_start) {{
+        first_start = event_start;
+    }}
+    if (event_end > last_end) {{
+        last_end = event_end;
+    }}
+    __state->report.add_completion("{module_name}", "FPGA", event_start, event_end, {sdfg.sdfg_id}, {state_id}, {state.node_id(sg.nodes()[0])});{print_str}
```

Comment:
Is it ok if this is called multiple time with the same module_name?

Reply:
You mean if the full kernel is invoked multiple times, or if there are multiple modules with the same name? I believe it will aggregate results reported for the same module (although that's currently based on a random node from the subgraph, rather than the name of the module itself, until I find a way to change this).

Reply:
I meant if the kernel is invoked multiple times (if multiple modules have the same name maybe it explodes earlier)

Reply:
That's fine, and an expected use case of the instrumentation flow :-)